moveos/moveos-stdlib/src/natives/moveos_stdlib/move_module.rs

```diff
@@ -0,0 +1,177 @@
+ **************************************************************************************************/
+#[derive(Clone, Debug)]
+pub struct ModuleNameInnerGasParameters {
+    pub base: InternalGas,
+    pub per_byte_in_str: InternalGasPerByte,
+}
+
+fn native_module_name_inner(
+    gas_params: &ModuleNameInnerGasParameters,
+    _context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    let byte_codes = pop_arg!(args, VectorRef);
+    let byte_codes_ref = byte_codes.as_bytes_ref();
+
+    let module = CompiledModule::deserialize(&byte_codes_ref)?;
+    let name = module.self_id().name().to_owned().into_string();
+    let cost = gas_params.base
+        + if gas_params.per_byte_in_str > 0.into() {
+            //TODO charge gas cost
+            gas_params.per_byte_in_str * NumBytes::new(name.len() as u64)
+        } else {
+            0.into()
+        };
+    let output = Struct::pack(vec![Value::vector_u8(name.as_bytes().to_vec())]);
+    let output_value = Value::struct_(output);
+    Ok(NativeResult::ok(cost, smallvec![output_value]))
+}
+
+/***************************************************************************************************
+ * native fun verify_modules_inner(
+ *      modules: &vector<vector<u8>>,
+ *      account_address: address
+ * ): vector<String>;
+ **************************************************************************************************/
+
+#[derive(Clone, Debug)]
+pub struct VerifyModulesGasParameters {
+    pub base: InternalGas,
+    pub per_byte: InternalGasPerByte,
+}
+
+fn native_verify_modules_inner(
+    gas_params: &VerifyModulesGasParameters,
+    context: &mut NativeContext,
+    _ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    let mut cost = gas_params.base;
+    let account_address = pop_arg!(args, AccountAddress);
+    let mut bundle = vec![];
+    for module in pop_arg!(args, Vec<Value>) {
+        let byte_codes = module.value_as::<Vec<u8>>()?;
+        cost += gas_params.per_byte * NumBytes::new(byte_codes.len() as u64);
+        bundle.push(byte_codes);
+    }
+    let compiled_modules = bundle
+        .iter()
+        .map(|b| CompiledModule::deserialize(b))
+        .collect::<PartialVMResult<Vec<CompiledModule>>>()?;
+    let module_context = context.extensions_mut().get_mut::<NativeModuleContext>();
+    let mut module_names = vec![];
+    for module in &compiled_modules {
+        if *module.self_id().address() != account_address {
+            return Err(PartialVMError::new(StatusCode::VERIFICATION_ERROR)
+                .with_message("Module address not match with signer".to_owned()));
+        }
+        let result = moveos_verifier::verifier::verify_module(module, module_context.resolver);
+        match result {
+            Ok(res) => {
+                if res {
+                    module_context.init_functions.push(module.self_id());
```

Comment:
the `verify_modules` function is not only used when publishing modules. Or we make the verify_modules to `friend`, and ensure it is only used in the `publish_modules` function.

Reply:
Or we put the init_function into TxContext in the publish_module function. The verify_module function needs to return if the module contains the init function.

Reply:
> > Or we put the init_function into TxContext in the publish_module function. The verify_module function needs to return if the module contains the init function.
> 
> Store init functions in TxContext.map ? The map use name of the value type as key, which may be not suitable to store init functions as we may dynamicly add more init functions with same types.

We need to define a struct and hold all init functions in a vector fields.